src/datasource.ts

```diff
@@ -0,0 +1,300 @@
+import {
+  ArrayVector,
+  DataFrame,
+  DataQueryRequest,
+  DataQueryResponse,
+  DataSourceApi,
+  DataSourceInstanceSettings,
+  Field,
+  FieldType,
+} from '@grafana/data';
+import { getBackendSrv } from '@grafana/runtime';
+import { lastValueFrom } from 'rxjs';
+import { CriblQuery, CriblDataSourceOptions } from './types';
+
+const QUERY_PAGE_SIZE = 1000;
+const CRIBL_TIME_FIELD = '_time';
+
+/**
+ * Cribl Search DataSource
+ */
+export class CriblDataSource extends DataSourceApi<CriblQuery, CriblDataSourceOptions> {
+  private authToken?: AuthToken;
+
+  constructor(private readonly instanceSettings: DataSourceInstanceSettings<CriblDataSourceOptions>) {
+    super(instanceSettings);
+  }
+
+  /**
+   * Run a query, loading the most recent results for the given savedQueryId
+   * @param options the options for this query
+   * @returns the DataQueryResponse containing DataFrame(s)
+   */
+  async query(options: DataQueryRequest<CriblQuery>): Promise<DataQueryResponse> {
+    // options.targets contains one or more queries, each of which results in a DataFrame
+    return Promise.all(options.targets.map((q) => this.processQuery(q))).then((data) => ({ data }));
+  }
+
+  /**
+   * Run a single query to produce a single DataFrame
+   * @param criblQuery the query to run
+   * @returns the DataFrame with the results
+   */
+  private async processQuery(criblQuery: CriblQuery): Promise<DataFrame> {
+    let fields: Record<string, Field> = {};
+    let eventCount = 0;
+    let totalEventCount: number | undefined = undefined;
+    let maxEventCount = criblQuery.maxResults ?? Number.MAX_VALUE;
+
+    // Load the search results, paging through until we've hit maxResults or read them all, whatever comes first
+    do {
+      const response = await this.getAuthorization() // refresh auth token as needed
+        .then((authorization) => lastValueFrom(getBackendSrv().fetch({
```

Comment:
nit: not sure how much I like this combination of await and promise-chaining via .then (btw: no .catch). Could we stick to await only and make this code look sequential?

Reply:
Yep, removed the chaining.  FWIW, the `await` here should have precluded the need for an explicit `catch` -- any error thrown at any point in the chain would end up percolating.  But I'm with you, don't want to use a style that isn't preferred by the team, so I changed it.